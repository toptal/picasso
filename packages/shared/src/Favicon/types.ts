export type EnvironmentType = 'development' | 'staging' | 'production'
