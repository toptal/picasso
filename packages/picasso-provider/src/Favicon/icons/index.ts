import type { EnvironmentType } from '../../types'
import * as temploy from './temploy'
import * as staging from './staging'
import * as production from './production'
import * as development from './development'

const getIcons = async (environment: EnvironmentType<'temploy'>) => {
  switch (environment) {
    case 'temploy':
      return temploy
    case 'staging':
      return staging
    case 'production':
      return production
    default:
      return development
  }
}

export { getIcons }
