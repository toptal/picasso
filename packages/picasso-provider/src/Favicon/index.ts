export { default, Favicon } from './Favicon'
export type { Props as FaviconProps } from './Favicon'
