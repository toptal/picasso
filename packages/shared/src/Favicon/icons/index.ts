import { EnvironmentType } from '../../Picasso'

const getIcons = (environment: EnvironmentType) => {
  if (environment === 'staging') {
    return import('./staging')
  }

  if (environment === 'production') {
    return import('./production')
  }

  return import('./development')
}

export { getIcons }
