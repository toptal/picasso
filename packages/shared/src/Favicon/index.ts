export { default } from './Favicon'
export * from './types'
