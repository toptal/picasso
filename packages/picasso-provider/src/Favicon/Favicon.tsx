import React, { useEffect } from 'react'
// we could use Page.Head instead, but it required PicassoProvider to be initialized
// so it led to cross dependencies and error
import { Helmet } from 'react-helmet-async'

import { useSafeState } from '../utils'
import { getIcons } from './icons'
import type { EnvironmentType } from '../types'
import { useAppConfig } from '../Picasso/RootContext'

export interface Props {
  /** Choose an icon color based on this variable */
  environment?: EnvironmentType<'test' | 'temploy'>
}

export const Favicon = ({ environment }: Props) => {
  const [icons, setIcons] = useSafeState<{
    icon16?: string
    icon32?: string
    icon180?: string
  }>({})

  const { environment: configEnvironment } = useAppConfig()

  const resolvedEnvironment = environment || configEnvironment

  useEffect(() => {
    if (resolvedEnvironment === 'test') {
      return
    }

    const loadIcons = async () => {
      try {
        const loadedIcons = await getIcons(
          resolvedEnvironment as EnvironmentType
        )

        setIcons(loadedIcons)
      } catch {
        console.error(
          'favicons were not loaded properly for environment' +
            resolvedEnvironment
        )
      }
    }

    loadIcons()
  }, [resolvedEnvironment, setIcons])

  if (resolvedEnvironment === 'test') {
    // do not load favicons in tests (e.g. in e2e)
    return null
  }

  const { icon16, icon32, icon180 } = icons

  // favicon.ico will be loaded automatically by browser from "public" folder
  // if it exists there. It's needed only for old browsers
  // we don't have to specify it in the head
  return (
    <Helmet>
      <link rel='apple-touch-icon' sizes='180x180' href={icon180} />
      <link rel='icon' type='image/png' sizes='32x32' href={icon32} />
      <link rel='icon' type='image/png' sizes='16x16' href={icon16} />
    </Helmet>
  )
}

Favicon.defaultProps = {}

Favicon.displayName = 'Favicon'

export default Favicon
