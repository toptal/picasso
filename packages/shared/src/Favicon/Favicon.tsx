import React, { useState, useEffect } from 'react'
// we could use Page.Head instead, but it required PicassoProvider to be initialized
// so it led to cross dependencies and error
import { Helmet } from 'react-helmet'

import { EnvironmentType } from './types'
import { getIcons } from './icons'

export interface Props {
  /** Choose an icon color based on this variable */
  environment?: EnvironmentType
}

export const Favicon = ({ environment }: Props) => {
  const [icons, setIcons] = useState<{
    icon16?: string
    icon32?: string
    icon180?: string
  }>({})

  useEffect(() => {
    async function loadIcons() {
      try {
        const icons = await getIcons(environment!)

        setIcons(icons)
      } catch {
        // eslint-disable-next-line no-console
        console.error(
          'favicons were not loaded properly for environment',
          environment
        )
      }
    }

    loadIcons()
  }, [])

  const { icon16, icon32, icon180 } = icons

  // favicon.ico will be loaded automatically by browser from "public" folder
  // if it exists there. It's needed only for old browsers
  // we don't have to specify it in the head
  return (
    <Helmet>
      <link rel='apple-touch-icon' sizes='180x180' href={icon180} />
      <link rel='icon' type='image/png' sizes='32x32' href={icon32} />
      <link rel='icon' type='image/png' sizes='16x16' href={icon16} />
    </Helmet>
  )
}

Favicon.defaultProps = {
  environment: 'development'
}

Favicon.displayName = 'Favicon'

export default Favicon
