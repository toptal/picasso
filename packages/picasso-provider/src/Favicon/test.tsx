import React from 'react'
import { render, waitFor } from '@toptal/picasso-test-utils'

import { Favicon } from './Favicon'

describe('Favicon', () => {
  it('renders', async () => {
    render(<Favicon />)

    await waitFor(() => {
      expect(document.querySelectorAll('head > link')).not.toHaveLength(0)
    })

    expect(document.querySelectorAll('head > link')).toMatchSnapshot()
  })

  it('renders with environment specified', async () => {
    render(<Favicon environment='staging' />)

    await waitFor(() => {
      expect(document.querySelectorAll('head > link')).not.toHaveLength(0)
    })

    expect(document.querySelectorAll('head > link')).toMatchSnapshot()
  })

  it('renders nothing in a test environment', async () => {
    const { container } = render(<Favicon environment='test' />)
    const picassoRoot = container.firstChild as Element

    expect(picassoRoot.children).toHaveLength(0)
  })
})
