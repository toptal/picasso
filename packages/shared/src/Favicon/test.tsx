import React from 'react'
import { render, waitForDomChange } from '@toptal/picasso/test-utils'

import Favicon from './Favicon'

test('default render', async () => {
  render(<Favicon />)

  await waitForDomChange()

  expect(document.querySelectorAll('head > link')).toMatchSnapshot()
})

test('with environment specified', async () => {
  render(<Favicon environment='staging' />)

  await waitForDomChange()

  expect(document.querySelectorAll('head > link')).toMatchSnapshot()
})

test('with test environment renders nothing', async () => {
  const { container } = render(<Favicon environment='test' />)
  const picassoRoot = container.firstChild as Element

  expect(picassoRoot.children.length).toBe(0)
})
