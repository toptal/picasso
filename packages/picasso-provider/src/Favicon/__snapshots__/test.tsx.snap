// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Favicon renders 1`] = `
NodeList [
  <link
    data-rh="true"
    href="test-file-stub"
    rel="apple-touch-icon"
    sizes="180x180"
  />,
  <link
    data-rh="true"
    href="test-file-stub"
    rel="icon"
    sizes="32x32"
    type="image/png"
  />,
  <link
    data-rh="true"
    href="test-file-stub"
    rel="icon"
    sizes="16x16"
    type="image/png"
  />,
]
`;

exports[`Favicon renders with environment specified 1`] = `
NodeList [
  <link
    data-rh="true"
    href="test-file-stub"
    rel="apple-touch-icon"
    sizes="180x180"
  />,
  <link
    data-rh="true"
    href="test-file-stub"
    rel="icon"
    sizes="32x32"
    type="image/png"
  />,
  <link
    data-rh="true"
    href="test-file-stub"
    rel="icon"
    sizes="16x16"
    type="image/png"
  />,
]
`;
